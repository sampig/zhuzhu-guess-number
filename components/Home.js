import React from 'react'
import { Alert, Button, StyleSheet, Text, View } from 'react-native'
import i18n from 'util/i18n-utils'
import useNumber from 'hooks/useNumber'
import GuessListMain from './GuessListMain'
import InputNumberForm from './InputNumberForm'

const styles = StyleSheet.create({
  containerButtons: {
    marginTop: 5,
    padding: 5,
    flexDirection: 'row',
    alignItems: 'center',
  },
  button: {
    marginLeft: 25,
    marginRight: 25,
    flex: 1,
  },
  containerList: {
    flex: 1,
    display: 'flex',
  },
})

/**
 * Component of Home page.
 * @param {object} navigation - the navigation object in route.
 * @returns {node}            - React node
 * @constructor
 */
function Home ({ navigation }) {
  const { gameParam, resetGame, tryGuess } = useNumber()

  /**
   * A function to open a to-do.
   */
  const openDetailPage = () => {
    navigation.navigate('Help')
  }

  /**
   * Restart game.
   */
  const restartGame = () => {
    if (!gameParam?.game?.number) {
      resetGame()
    } else {
      Alert.alert(
        i18n.t('restartGameTitle'),
        i18n.t('restartGameText'),
        [
          {
            text: i18n.t('cancelButton'),
            style: 'cancel',
          },
          {
            text: i18n.t('startButton'),
            onPress: () => {
              resetGame()
            },
          },
        ],
        {
          cancelable: true,
        },
      )
    }
  }

  /**
   * Buttons at the bottom of the page.
   * @returns {node} - React node of buttons
   */
  const buttonsNav = () => {
    return (
      <View style={styles.containerButtons}>
        <View style={styles.button}>
          <Button
            title={i18n.t('help')}
            onPress={openDetailPage}
          />
        </View>
        <View style={styles.button}>
          <Button
            title={i18n.t('newGame')}
            onPress={restartGame}
          />
        </View>
      </View>
    )
  }

  /**
   * Get game status.
   * @returns {string} - game status
   */
  const gameStatus = () => {
    if (!gameParam?.game?.number) {
      return i18n.t('gameStatusNotStart')
    } else if (gameParam.isWin() || gameParam.isGameOver()) {
      return i18n.t('gameStatusOver')
    }
    return i18n.t('gameStatusRunning', { times: 10 - gameParam.guessList.length })
  }

  return (
    <>
      {buttonsNav()}
      <Text>{i18n.t('gameStatus') + ' ' + gameStatus()}</Text>
      <GuessListMain gameParam={gameParam} />
      <InputNumberForm gameParam={gameParam} tryGuess={tryGuess} />
    </>
  )
}

export default Home
