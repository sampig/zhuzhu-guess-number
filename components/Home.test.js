import React from 'react'
import renderer from 'react-test-renderer'
import Home from './Home'

const navigation = { navigate: jest.fn() }
jest.mock('hooks/useNumber', () => ({
  __esModule: true,
  default: () => ({
    gameParam: {
      game: {
        number: [1, 2, 3, 4],
        times: 0,
      },
      guessList: [],
      isWin: jest.fn(),
      isGameOver: jest.fn(),
    },
    resetGame: jest.fn(),
    tryGuess: jest.fn(),
  }),
}))

describe('The Home component', () => {
  it('renders without crashing', () => {
    const component = renderer.create(<Home navigation={navigation} />)
    const render = component.toJSON()
    expect(render).toMatchSnapshot()
  })
})
