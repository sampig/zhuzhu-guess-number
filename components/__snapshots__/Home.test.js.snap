// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Home component renders without crashing 1`] = `
Array [
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginTop": 5,
        "padding": 5,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "marginLeft": 25,
          "marginRight": 25,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Help
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "marginLeft": 25,
          "marginRight": 25,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            New Game
          </Text>
        </View>
      </View>
    </View>
  </View>,
  <Text>
    Game status: Game started. You have 10 chances left.
  </Text>,
  <View
    style={
      Object {
        "borderRadius": 5,
        "borderWidth": 1,
        "flex": 1,
        "margin": 5,
        "padding": 15,
      }
    }
  >
    <Text>
      Your guess:
    </Text>
    <RCTScrollView
      data={Array []}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "display": "flex",
          "flex": 1,
        }
      }
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View />
    </RCTScrollView>
  </View>,
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "flexDirection": "row",
        "paddingRight": 6,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      editable={true}
      onChangeText={[Function]}
      placeholder="Enter a 4-digit number"
      rejectResponderTermination={true}
      style={
        Object {
          "borderRadius": 5,
          "borderWidth": 2,
          "flex": 1,
          "fontSize": 20,
          "margin": 6,
          "padding": 6,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
            Object {},
          ]
        }
      >
        <Text
          disabled={true}
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#cdcdcd",
              },
            ]
          }
        >
          Guess
        </Text>
      </View>
    </View>
  </View>,
]
`;
